packages/@aws-cdk/aws-iam/lib/private/merge-statements.ts

```diff
@@ -59,34 +89,41 @@ export function mergeStatements(statements: StatementSchema[]): StatementSchema[
  * - From their Action, Resource and Principal sets, 2 are subsets of each other
  *   (empty sets are fine).
  */
-function tryMerge(a: ComparableStatement, b: ComparableStatement): ComparableStatement | undefined {
+function tryMerge(a: ComparableStatement, b: ComparableStatement, limitSize: boolean): ComparableStatement | undefined {
   // Effects must be the same
-  if (a.effect !== b.effect) { return; }
+  if (a.statement.effect !== b.statement.effect) { return; }
   // We don't merge Sids (for now)
-  if (a.sid || b.sid) { return; }
+  if (a.statement.sid || b.statement.sid) { return; }
 
   if (a.conditionString !== b.conditionString) { return; }
-  if (!setEqual(a.notAction, b.notAction) || !setEqual(a.notResource, b.notResource) || !setEqual(a.notPrincipal, b.notPrincipal)) { return; }
+  if (
+    !setEqual(a.statement.notActions, b.statement.notActions) ||
+    !setEqual(a.statement.notResources, b.statement.notResources) ||
+    !setEqualPrincipals(a.statement.notPrincipals, b.statement.notPrincipals)
+  ) {
+    return;
+  }
 
   // We can merge these statements if 2 out of the 3 sets of Action, Resource, Principal
   // are the same.
-  const setsEqual = (setEqual(a.action, b.action) ? 1 : 0) +
-    (setEqual(a.resource, b.resource) ? 1 : 0) +
-    (setEqual(a.principal, b.principal) ? 1 : 0);
+  const setsEqual = (setEqual(a.statement.actions, b.statement.actions) ? 1 : 0) +
+    (setEqual(a.statement.resources, b.statement.resources) ? 1 : 0) +
+    (setEqualPrincipals(a.statement.principals, b.statement.principals) ? 1 : 0);
 
   if (setsEqual < 2 || unmergeablePrincipals(a, b)) { return; }
 
+  const combined = a.statement.copy({
+    actions: setMerge(a.statement.actions, b.statement.actions),
+    resources: setMerge(a.statement.resources, b.statement.resources),
+    principals: setMergePrincipals(a.statement.principals, b.statement.principals),
+  });
+
+  if (limitSize && combined._estimateSize() > MAX_MERGE_SIZE) { return undefined; }
```

Comment:
I am a bit weary of relying on estimate sizes, but I don't see a better way here...